isr_tables: Support hardware interrupt vector table-only configuration.

stephanosio opened the pull request:
The existing isr_tables implementation does not allow enabling only
hardware interrupt vector table without software isr table.

This commit ensures that CONFIG_GEN_IRQ_VECTOR_TABLE can be used
without setting CONFIG_GEN_SW_ISR_TABLE.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

NOTE: This is necessary to support the Cortex-R cores (Cortex-R4 and -R5) with the Vectored Interrupt Controller (VIC) port that allows direct interrupt vectoring.

stephanosio committed:
isr_tables: Support hardware interrupt vector table-only configuration.

The existing isr_tables implementation does not allow enabling only
hardware interrupt vector table without software isr table.

This commit ensures that CONFIG_GEN_IRQ_VECTOR_TABLE can be used
without setting CONFIG_GEN_SW_ISR_TABLE.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

SebastianBoe commented on arch/common/isr_tables.c:
```diff
@@ -29,18 +29,21 @@ Z_GENERIC_SECTION(.irq_info) struct int_list_header _iheader = {
 	[0 ...(IRQ_TABLE_SIZE - 1)] = (u32_t)&_isr_wrapper,
+#else
+	[0 ...(IRQ_TABLE_SIZE - 1)] = (u32_t)&z_irq_spurious,
+#endif
```
Put _isr_wrapper and z_irq_spurious in it's own define and ifdef just that.

SebastianBoe commented:
I'm not familiar enough with the irq handling to review, removing myself as reviewer.

jhedberg commented:
@stephanosio @andrewboie is this ready to go in, or you're waiting for more reviews? Btw, I just triggered a Shippable re-run since it was a month ago when it last happened.

stephanosio commented:
> @stephanosio @andrewboie is this ready to go in, or you're waiting for more reviews? Btw, I just triggered a Shippable re-run since it was a month ago when it last happened.

Given this patch's potential to break a lot of things, I suppose more reviews can't hurt before merging.

This is not terribly urgent as it will take some time for me to upstream the TI Hercules SoC port and that will likely be the only user of it.

@andyross @ioannisg your insight on this would be greatly appreciated.

ioannisg commented:
> @andyross @ioannisg your insight on this would be greatly appreciated.

I will take a look as soon as possible

ioannisg commented on arch/common/isr_tables.c:
```diff
@@ -29,20 +29,29 @@ Z_GENERIC_SECTION(.irq_info) struct int_list_header _iheader = {
+ */
+#ifdef CONFIG_GEN_SW_ISR_TABLE
+#define IRQ_VECTOR_TABLE_DEFAULT_ISR	_isr_wrapper
+#else
```
Could you also add a comment here?

ioannisg replied:
how about an inline comment here at the #else? This seems to be the differentiation brought in this patch

stephanosio replied:
I am not sure if I understand what you mean by "inline comment" here.

ioannisg replied:
Well I'd like a comment as the one in lines 41-44, for the #else case, i.e. when only the vector tables are used.

stephanosio replied:
I thought it was quite self-explanatory -- that unconnected IRQs should have `z_irq_spurious`, as you would in `_sw_isr_table`.

I will add a comment explaining this anyway, since this might not be as obvious to the person reading this for the first time.

ioannisg replied:
yes, please

stephanosio replied:
```
/* When both the IRQ vector table and the software ISR table are used, populate
 * the IRQ vector table with the common software ISR by default, such that all
 * indirect interrupt vectors are handled using the software ISR table;
 * otherwise, populate the IRQ vector table with z_irq_spurious so that all
 * un-connected IRQ vectors end up in the spurious IRQ handler.
 */
```

ioannisg reviewed (commented):
Looks fine, some requests for comments' enhancements only.

stephanosio commented:
> Looks fine, some requests for comments' enhancements only.

@ioannisg Thanks for reviewing. I have addressed your comments and pushed a new commit.

ioannisg commented:
> Given this patch's potential to break a lot of things, I suppose more reviews can't hurt before merging.

Let's merge this one - we are early in the 2.3 release cycle, so hopefully, there is enough time to detect and fix any bad effects of this PR
